compiler/src/iree/compiler/Dialect/HAL/Utils/ExternBuildingUtils.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_DIALECT_HAL_UTILS_EXTERN_BUILDING_UTILS_H_
+#define IREE_COMPILER_DIALECT_HAL_UTILS_EXTERN_BUILDING_UTILS_H_
+
+#include "iree/compiler/Dialect/HAL/Utils/ExternBuildingUtils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "mlir/Dialect/PDLInterp/IR/PDLInterp.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Location.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+namespace {
+
+/// Helper to build a hal.dispatch.extern op with the given arguments. This
+/// returns the block arguments of the workgroup count region and the extern
+/// op. Note that the workgroup count region will not include the terminator
+/// and that is left up to the user to properly populate.
+static FailureOr<Operation *>
+createDispatchExtern(PatternRewriter &rewriter, ValueRange workload,
+                     TypeRange resultTypes, ValueRange resultDims,
+                     ValueRange arguments, ValueRange argumentDims,
+                     DenseI64ArrayAttr tiedOperands, DictionaryAttr attrDict) {
+  Location rootLoc = (*arguments.begin()).getLoc();
```

Comment:
having not used PDL, is there no better location that could be used? seems like this would lead to very confusing IR (the location of argument 0 could be an arith.constant or something random, and seeing a big dispatch show up with the source loc of a constant is odd :)

in your pattern_module.mlir I'd expect the loc of the %mul here - can you get its location and pass that in as the first call arg?

Reply:
Good question; I'm hoping this is improved when I switch this out for a normal op builder (given that PDL automatically handles location information in that case, and I would hope it would use the root of the pdl.rewrite).

Reply:
having some difficulties with PDL outlined here: https://github.com/llvm/llvm-project/issues/69888

Reply:
I tried using the pdl operation building functionality directly, but it doesn't seem to want to construct the op properly, e.g.
```
%0 = "hal.dispatch.extern"(<<UNKNOWN SSA VALUE>>, <<UNKNOWN SSA VALUE>>, <<UNKNOWN SSA VALUE>>, <<UNKNOWN SSA VALUE>>, <<UNKNOWN SSA VALUE>>, <<UNKNOWN SSA VALUE>>) {bindings = [#hal.interface.binding<0, 0>, #hal.interface.binding<0, 1>, #hal.interface.binding<0, 2>], ... other attributes
```
I don't know if it's related to the removed default op builder or some operand segment length issue, but I'm going to stick to the wrapper around the builder for now and just pass in the replacement target to get the location information (because it works :P).